File: samples/Windows/java/NonTableManagerSamples/OpenPositionNetting/src/ResponseListener.java
package openpositionnetting;

import java.util.List;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

import com.fxcore2.*;
import common.OrderMonitorNetting;

public class ResponseListener implements IO2GResponseListener {

    private O2GSession mSession;
    private String mRequestID;
    private O2GResponse mResponse;
    private Semaphore mSemaphore;
    private OrderMonitorNetting mOrderMonitor;
    private O2GTradesTableResponseReader mTradesTable;

    // ctor
    public ResponseListener(O2GSession session) {
        mSession = session;
        mRequestID = "";
        mResponse = null;
        mSemaphore = new Semaphore(0);
        mOrderMonitor = null;
        mTradesTable = null;
    }

    public void setRequestID(String sRequestID) {
        mResponse = null;
        mRequestID = sRequestID;
    }

    public boolean waitEvents() throws Exception {
        return mSemaphore.tryAcquire(30, TimeUnit.SECONDS);
    }

    public O2GResponse getResponse() {
        return mResponse;
    }
    
    public void setTradesTable(O2GTradesTableResponseReader tradesTable) {
        mTradesTable = tradesTable;
    }

    public void onRequestCompleted(String sRequestID, O2GResponse response) {
        if (mRequestID.equals(response.getRequestId())) {
            mResponse = response;
            mSemaphore.release();
        }
        // real order execution is processed by Order monitor
    }

    public void onRequestFailed(String sRequestID, String sError) {
        if (mRequestID.equals(sRequestID)) {
            System.out.println("Request failed: " + sError);
            mSemaphore.release();
        }
    }

    public void onTablesUpdates(O2GResponse response) {
        O2GResponseReaderFactory factory = mSession.getResponseReaderFactory();
        if (factory != null) {
            O2GTablesUpdatesReader reader = factory.createTablesUpdatesReader(response);
            for (int ii = 0; ii < reader.size(); ii++) {
                switch (reader.getUpdateTable(ii)) {
                case ACCOUNTS:
                    O2GAccountRow account = reader.getAccountRow(ii);
                    // Show balance updates
                    System.out.println("Balance: " + account.getBalance());
                    break;
                case ORDERS:
                    O2GOrderRow order = reader.getOrderRow(ii);
                    switch (reader.getUpdateType(ii)) {
                    case INSERT:
                        if ((OrderMonitorNetting.isClosingOrder(order) || OrderMonitorNetting.isOpeningOrder(order)) &&
                                mOrderMonitor == null)
                        {
                            System.out.println(String.format("The order has been added. Order ID: %s, Rate: %s, Time In Force: %s",
                                    order.getOrderID(),
                                    order.getRate(),
                                    order.getTimeInForce()));
                            O2GTradeRow tradeRow = null;
                            String sTradeID = order.getTradeID();
                            if (mTradesTable != null) {
                                for (int j = 0; j < mTradesTable.size(); j++) {
                                    if (sTradeID.equals(mTradesTable.getRow(j).getTradeID())) {
                                        tradeRow = mTradesTable.getRow(j);
                                        break;
                                    }
                                }
                            }
                            if (tradeRow == null) {
                                mOrderMonitor = new OrderMonitorNetting(order);
                            } else {
                                mOrderMonitor = new OrderMonitorNetting(order, tradeRow.getAmount());
                            }
                        }
                        break;
                    case DELETE:
                        if (mOrderMonitor != null) {
                            System.out.println(String.format("The order has been deleted. Order ID: %s",
                                    order.getOrderID()));
                            mOrderMonitor.onOrderDeleted(order);
                            if (mOrderMonitor.isOrderCompleted()) {
                                printResult();
                                mSemaphore.release();
                            }
                        }
                        break;
                    }
                    break;
                case TRADES:
                    O2GTradeRow trade = reader.getTradeRow(ii);
                    switch (reader.getUpdateType(ii)) {
                        case INSERT:
                            if (mOrderMonitor != null) {
                                mOrderMonitor.onTradeAdded(trade);
                                if (mOrderMonitor.isOrderCompleted()) {
                                    printResult();
                                    mSemaphore.release();
                                }
                            }
                            break;
                        case UPDATE:
                            if (mOrderMonitor != null)
                            {
                                mOrderMonitor.onTradeUpdated(trade);
                                if (mOrderMonitor.isOrderCompleted()) {
                                    printResult();
                                    mSemaphore.release();
                                }
                            }
                            break;
                    }
                    break;
                case CLOSED_TRADES:
                    O2GClosedTradeRow closedTrade = reader.getClosedTradeRow(ii);
                    if (reader.getUpdateType(ii) == O2GTableUpdateType.INSERT) {
                        if (mOrderMonitor != null) {
                            mOrderMonitor.onClosedTradeAdded(closedTrade);
                            if (mOrderMonitor.isOrderCompleted()) {
                                printResult();
                                mSemaphore.release();
                            }
                        }
                    }
                    break;
                case MESSAGES:
                    O2GMessageRow message = reader.getMessageRow(ii);
                    if (reader.getUpdateType(ii) == O2GTableUpdateType.INSERT) {
                        if (mOrderMonitor != null) {
                            mOrderMonitor.onMessageAdded(message);
                            if (mOrderMonitor.isOrderCompleted()) {
                                printResult();
                                mSemaphore.release();
                            }
                        }
                    }
                    break;
                }
            }
        }
    }

    private void printResult() {
        if (mOrderMonitor != null) {
            OrderMonitorNetting.ExecutionResult result = mOrderMonitor.getResult();
            List<O2GTradeRow> trades;
            List<O2GTradeRow> updatedTrades;
            List<O2GClosedTradeRow> closedTrades;
            O2GOrderRow order = mOrderMonitor.getOrder();
            String sOrderID = order.getOrderID();
            trades = mOrderMonitor.getTrades();
            updatedTrades = mOrderMonitor.getUpdatedTrades();
            closedTrades = mOrderMonitor.getClosedTrades();

            switch (result) {
            case Canceled:
                if (trades.size() > 0) {
                    printTrades(trades, sOrderID);
                    printUpdatedTrades(updatedTrades, sOrderID);
                    printClosedTrades(closedTrades, sOrderID);
                    System.out.println(String.format("A part of the order has been canceled. Amount = %s",
                            mOrderMonitor.getRejectAmount()));
                } else {
                    System.out.println(String.format(
                            "The order: OrderID = %s  has been canceled.",
                            sOrderID));
                    System.out.println(String.format(
                            "The cancel amount = %s.",
                            mOrderMonitor.getRejectAmount()));
                }
                break;
            case FullyRejected:
                System.out.println(String.format(
                        "The order has been rejected. OrderID = %s",
                        sOrderID));
                System.out.println(String.format("The rejected amount = %s",
                        mOrderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s",
                        mOrderMonitor.getRejectMessage()));
                break;
            case PartialRejected:
                printTrades(trades, sOrderID);
                printUpdatedTrades(updatedTrades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                System.out.println(String.format(
                        "A part of the order has been rejected. Amount = %s",
                        mOrderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s ",
                        mOrderMonitor.getRejectMessage()));
                break;
            case Executed:
                printTrades(trades, sOrderID);
                printUpdatedTrades(updatedTrades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                break;
            }
        }
    }

    private void printTrades(List<O2GTradeRow> trades, String sOrderID) {
        if (trades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been opened:",
                sOrderID));

        for (int i = 0; i < trades.size(); i++) {
            O2GTradeRow trade = trades.get(i);
            String sTradeID = trade.getTradeID();
            int iAmount = trade.getAmount();
            double dRate = trade.getOpenRate();
            System.out.println(String.format(
                    "Trade ID: %s; Amount: %s; Rate: %s",
                    sTradeID, iAmount, dRate));
        }
    }
    
    private void printUpdatedTrades(List<O2GTradeRow> updatedTrades, String sOrderID) {
        if (updatedTrades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been updated:",
                sOrderID));

        for (int i = 0; i < updatedTrades.size(); i++) {
            O2GTradeRow trade = updatedTrades.get(i);
            String sTradeID = trade.getTradeID();
            int iAmount = trade.getAmount();
            double dRate = trade.getOpenRate();
            System.out.println(String.format(
                    "Trade ID: %s; Amount: %s; Rate: %s",
                    sTradeID, iAmount, dRate));
        }
    }

    private void printClosedTrades(List<O2GClosedTradeRow> closedTrades,
            String sOrderID) {
        if (closedTrades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been closed: ",
                sOrderID));

        for (int i = 0; i < closedTrades.size(); i++) {
            O2GClosedTradeRow closedTrade = closedTrades.get(i);
            String sTradeID = closedTrade.getTradeID();
            int iAmount = closedTrade.getAmount();
            double dRate = closedTrade.getCloseRate();
            System.out.println(String.format(
                    "Closed Trade ID: %s; Amount: %s; Closed Rate: %s",
                    sTradeID, iAmount, dRate));
        }
    }
}

File: samples/macOS/cs/netcore1.0/NonTableManagerSamples/OpenPosition/src/ResponseListener.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using System.Threading;
using fxcore2;
using OrdersMonitor;

namespace OpenPosition
{
    class ResponseListener : IO2GResponseListener
    {
        private O2GSession mSession;
        private string mRequestID;
        private O2GResponse mResponse;
        private EventWaitHandle mSyncResponseEvent;
        private OrderMonitor mOrderMonitor;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="session"></param>
        public ResponseListener(O2GSession session)
        {
            mRequestID = string.Empty;
            mResponse = null;
            mOrderMonitor = null;
            mSyncResponseEvent = new EventWaitHandle(false, EventResetMode.AutoReset);
            mSession = session;            
        }

        public void SetRequestID(string sRequestID)
        {
            mResponse = null;
            mRequestID = sRequestID;
        }

        public bool WaitEvents()
        {
            return mSyncResponseEvent.WaitOne(30000);
        }

        public O2GResponse GetResponse()
        {
            return mResponse;
        }

        #region IO2GResponseListener Members

        public void onRequestCompleted(string sRequestId, O2GResponse response)
        {
            if (mRequestID.Equals(response.RequestID))
            {
                mResponse = response;
                if (response.Type != O2GResponseType.CreateOrderResponse)
                {
                    mSyncResponseEvent.Set();
                }
            }
            //real order execution is processed by Order monitor
        }

        public void onRequestFailed(string sRequestID, string sError)
        {
            if (mRequestID.Equals(sRequestID))
            {
                Console.WriteLine("Request failed: " + sError);
                mSyncResponseEvent.Set();
            }
        }

        public void onTablesUpdates(O2GResponse data)
        {
            O2GResponseReaderFactory factory = mSession.getResponseReaderFactory();
            if (factory != null)
            {
                O2GTablesUpdatesReader reader = factory.createTablesUpdatesReader(data);
                for (int ii = 0; ii < reader.Count; ii++)
                {
                    switch (reader.getUpdateTable(ii))
                    {
                        case O2GTableType.Accounts:
                            O2GAccountRow account = reader.getAccountRow(ii);
                            //Show balance updates
                            Console.WriteLine("Balance: {0}", account.Balance);
                            break;
                        case O2GTableType.Orders:
                            O2GOrderRow order = reader.getOrderRow(ii);
                            switch (reader.getUpdateType(ii))
                            {
                                case O2GTableUpdateType.Insert:
                                    if ((OrderMonitor.IsClosingOrder(order) || OrderMonitor.IsOpeningOrder(order)) &&
                                            mOrderMonitor == null)
                                    {
                                        Console.WriteLine("The order has been added. Order ID: {0}, Rate: {1}, Time In Force: {2}",
                                            order.OrderID,
                                            order.Rate,
                                            order.TimeInForce);
                                        mOrderMonitor = new OrderMonitor(order);
                                    }
                                    break;
                                case O2GTableUpdateType.Delete:
                                    if (mOrderMonitor != null)
                                    {
                                        Console.WriteLine("The order has been deleted. Order ID: {0}", order.OrderID);
                                        mOrderMonitor.OnOrderDeleted(order);
                                        if (mOrderMonitor.IsOrderCompleted)
                                        {
                                            PrintResult();
                                            mSyncResponseEvent.Set();
                                        }
                                    }
                                    break;
                            }
                            break;
                        case O2GTableType.Trades:
                            O2GTradeRow trade = reader.getTradeRow(ii);
                            if (reader.getUpdateType(ii) == O2GTableUpdateType.Insert)
                            {
                                if (mOrderMonitor != null)
                                {
                                    mOrderMonitor.OnTradeAdded(trade);
                                    if (mOrderMonitor.IsOrderCompleted)
                                    {
                                        PrintResult();
                                        mSyncResponseEvent.Set();
                                    }
                                }
                            }
                            break;
                        case O2GTableType.ClosedTrades:
                            O2GClosedTradeRow closedTrade = reader.getClosedTradeRow(ii);
                            if (reader.getUpdateType(ii) == O2GTableUpdateType.Insert)
                            {
                                if (mOrderMonitor != null)
                                {
                                    mOrderMonitor.OnClosedTradeAdded(closedTrade);
                                    if (mOrderMonitor.IsOrderCompleted)
                                    {
                                        PrintResult();
                                        mSyncResponseEvent.Set();
                                    }
                                }
                            }
                            break;
                        case O2GTableType.Messages:
                            O2GMessageRow message = reader.getMessageRow(ii);
                            if (reader.getUpdateType(ii) == O2GTableUpdateType.Insert)
                            {
                                if (mOrderMonitor != null)
                                {
                                    mOrderMonitor.OnMessageAdded(message);
                                    if (mOrderMonitor.IsOrderCompleted)
                                    {
                                        PrintResult();
                                        mSyncResponseEvent.Set();
                                    }
                                }
                            }
                            break;
                    }
                }
            }
        }

        #endregion

        private void PrintResult()
        {
            if (mOrderMonitor != null)
            {
                OrderMonitor.ExecutionResult result = mOrderMonitor.Result;
                ReadOnlyCollection<O2GTradeRow> trades;
                ReadOnlyCollection<O2GClosedTradeRow> closedTrades;
                O2GOrderRow order = mOrderMonitor.Order;
                string sOrderID = order.OrderID;
                trades = mOrderMonitor.Trades;
                closedTrades = mOrderMonitor.ClosedTrades;

                switch (result)
                {
                    case OrderMonitor.ExecutionResult.Canceled:
                        if (trades.Count > 0)
                        {
                            PrintTrades(trades, sOrderID);
                            PrintClosedTrades(closedTrades, sOrderID);
                            Console.WriteLine("A part of the order has been canceled. Amount = {0}", mOrderMonitor.RejectAmount);
                        }
                        else
                        {
                            Console.WriteLine("The order: OrderID = {0}  has been canceled.", sOrderID);
                            Console.WriteLine("The cancel amount = {0}.", mOrderMonitor.RejectAmount);
                        }
                        break;
                    case OrderMonitor.ExecutionResult.FullyRejected:
                        Console.WriteLine("The order has been rejected. OrderID = {0}", sOrderID);
                        Console.WriteLine("The rejected amount = {0}", mOrderMonitor.RejectAmount);
                        Console.WriteLine("Rejection cause: {0}", mOrderMonitor.RejectMessage);
                        break;
                    case OrderMonitor.ExecutionResult.PartialRejected:
                        PrintTrades(trades, sOrderID);
                        PrintClosedTrades(closedTrades, sOrderID);
                        Console.WriteLine("A part of the order has been rejected. Amount = {0}", mOrderMonitor.RejectAmount);
                        Console.WriteLine("Rejection cause: {0} ", mOrderMonitor.RejectMessage);
                        break;
                    case OrderMonitor.ExecutionResult.Executed:
                        PrintTrades(trades, sOrderID);
                        PrintClosedTrades(closedTrades, sOrderID);
                        break;
                }
            }
        }

        private void PrintTrades(ReadOnlyCollection<O2GTradeRow> trades, string sOrderID)
        {
            if (trades.Count == 0)
                return;
            Console.WriteLine("For the order: OrderID = {0} the following positions have been opened:", sOrderID);

            for (int i = 0; i < trades.Count; i++)
            {
                O2GTradeRow trade = trades[i];
                string sTradeID = trade.TradeID;
                int iAmount = trade.Amount;
                double dRate = trade.OpenRate;
                Console.WriteLine("Trade ID: {0}; Amount: {1}; Rate: {2}", sTradeID, iAmount, dRate);
            }
        }

        private void PrintClosedTrades(ReadOnlyCollection<O2GClosedTradeRow> closedTrades, string sOrderID)
        {
            if (closedTrades.Count == 0)
                return;
            Console.WriteLine("For the order: OrderID = {0} the following positions have been closed: ", sOrderID);

            for (int i = 0; i < closedTrades.Count; i++)
            {
                O2GClosedTradeRow closedTrade = closedTrades[i];
                string sTradeID = closedTrade.TradeID;
                int iAmount = closedTrade.Amount;
                double dRate = closedTrade.CloseRate;
                Console.WriteLine("Closed Trade ID: {0}; Amount: {1}; Closed Rate: {2}", sTradeID, iAmount, dRate);
            }
        }
    }
}

